File: Ceasar.java
/**!
 * Ceasar Cipher manipulation
 * Purpose: Generate a cipher based on a key utilizing shifting techniques
 * @Author: Xinthral
 * Date: 9/18/18
 */
import java.util.Arrays;

public class Ceasar {
  // CLASS SCOPE VARIABLES
  public static final char[] letters = {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
          'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'};
  public char[][] outputMatrix = new char[26][26];
  public char[][] shadowMatrix = new char[26][26];
  public char[] keyArray;

  // CLASS METHODS
  private int getIndex(char letter){
    /**!
     * Locates the index of the given letter in the alphabet
     * @param letter: The letter in which you wish to locate the index of
     * @return Integer representing the index of the given letter.
     */
    // Locate index of input code
    int indexNumber = 0;
    for (int x = 0; x < letters.length; x++) {
      if (letters[x] == letter){
        indexNumber = x;
      }
    }
    return indexNumber;
  }
  private void generateCipherMatrix(char code){
    /**!
     * Generates the Matrix with a given starting position
     * @param code : The first letter that the matrix starts at.
     */
    int startIndex = getIndex(code);

    for ( int i = 0; i < 26; i++) {
      char[] shiftedChars = new char[26];
      System.arraycopy(letters, startIndex, shiftedChars,0, letters.length - startIndex);
      System.arraycopy(letters, 0, shiftedChars, (letters.length - startIndex), letters.length - (letters.length - startIndex));
      outputMatrix[i] = shiftedChars;
      startIndex = (startIndex + 1) % 26;
      Arrays.fill(shadowMatrix[i], ' ');
    }
  }

  String encode(String inputText){
    /**!
     * Takes in a string to be encoded with the ciphered matrix
     * @param inputText: Given string you wish to be encoded.
     * @return The cipher encoded string.
     */

    char[] inputArray = inputText.toCharArray();
    String outputString = "";

    int keyIndex = 0;
    for(int i = 0; i < inputArray.length; i++) {
      if (inputArray[i] != ' ') {
        int firstIndex = getIndex(keyArray[keyIndex]);
        int secondIndex = getIndex(inputArray[i]);
        outputString += outputMatrix[firstIndex][secondIndex];
        shadowMatrix[firstIndex][secondIndex] = inputArray[i];
      } else { outputString += " "; }
      keyIndex = (keyIndex + 1) % keyArray.length;
    }
    return outputString;
  }

  String decode(String inputText){
    /**!
     * Takes in a string to be decoded with the ciphered matrix
     * @param inputText: Given string you wish to be decoded.
     * @return The cipher decoded string.
     */
    char[] inputArray = inputText.toCharArray();
    String outputString = "";

    int keyIndex = 0;
    for(int i = 0; i < inputArray.length; i++) {
      if (inputArray[i] != ' ') {
        char[] temp = outputMatrix[getIndex(keyArray[keyIndex])];
        for (int j = 0; j < temp.length; j++) {
          if (temp[j] == inputArray[i]) {
            outputString += letters[j];
          }
        }
      } else { outputString += " "; }
      keyIndex = (keyIndex + 1) % keyArray.length;
    }
    return outputString;
  }

  public void displayMatrix(boolean blackOut) {
    /**
     * Displays the matrix in an easy to read format.
     * @param hidden: displays the matrix without erroneous data
     *              ** WARNING ** - If true, can only be ran after a message
     *              has been decoded.
     */
    String[] matrixString = new String[outputMatrix.length];
    String[] shadowString = new String[outputMatrix.length];
    String header = "[ ]  ";
    Arrays.fill(matrixString, "");
    Arrays.fill(shadowString, "");

    for (int i = 0; i < outputMatrix.length; i++) {
      matrixString[i] += String.format("[%s] ", letters[i]);
      shadowString[i] += String.format("[%s] ", letters[i]);
      for (int j = 0; j < outputMatrix[i].length; j++) {
        matrixString[i] += String.format("| %s ", outputMatrix[i][j]);
        shadowString[i] += String.format("| %s ", shadowMatrix[i][j]);
      }
      matrixString[i] += "|";
      shadowString[i] += "|";
    }

    for (int i = 0; i < letters.length; i++) {
      header += String.format("[%c] ", letters[i]);
    }

    System.out.println(header);
    for (int i = 0; i < matrixString.length; i++) {
      System.out.println(blackOut ? shadowString[i] : matrixString[i]);
    }
  }

  public Ceasar(char code, String key) {
    /**!
     * Constructor method to initialize the ciphered matrix
     * @param code: The starting letter for the matrix.
     * @param key:  A given string that can be used to encrypt
     *              and decrypt a given message.
     */
    keyArray = key.toCharArray();
    generateCipherMatrix(code);
  }

  public static void main(String[] args) {
    char code = 'H';
    String key = "BABBAGE";
    String msg = "HAPPY BIRTHDAY";
    // boolean hidden = false;
    Ceasar self = new Ceasar(code, key);
    
    String response1 = self.encode(msg);
    assert "PHXXF MQYBPKNJ".equals(response1);
    String response2 = self.decode(response1);
    assert "HAPPY BIRTHDAY".equals(response2);

    // Optional Method for viewing the cipher matrix based on last item decoded
    // self.displayMatrix(false);
    String output = String.format("Input:     %s\nEncrypted: %s\nDecrypted: %s", msg, response1, response2);
    System.out.println(output);
  }
}

File: ceasar.rs
/* 
Caesar Cipher manipulation
Purpose: Generate a cipher based on a shift key (integer offset)
and integrating a salted sequence
Author: Xinthral
Date: 10/01/25
*/
pub struct Ceasar {
  shift_key: i32,
  mask: Vec<char>,
  letters: Vec<char>,
  matrix: Vec<Vec<char>>,
}

impl Ceasar {
  /// Constructor: creates a new Caesar cipher with given shift and salt
  pub fn new(key: char, salt: &str) -> Ceasar {
    let letters: Vec<char> = ('A'..='Z').collect();
    let matrix: Vec<Vec<char>> = generate_matrix(&key);
    let key_index: i32 = letters.iter().position(|&c| c == key).unwrap_or(0) as i32;
    Ceasar {
      shift_key: key_index.rem_euclid(26),
      mask: salt.to_uppercase().chars().collect(),
      letters,
      matrix,
    }
  }

  /// Finds the index of a letter in the alphabet
  fn get_index(&self, letter: char) -> Option<usize> {
    self.letters.iter().position(|&c| c == letter)
  }

  /// Shifts a character by the given amount
  fn shift_char(&self, c: char, shift: i32) -> char {
    if c == ' ' { return ' '; }
    match self.get_index(c) {
      Some(index) => {
        let new_index = (index as i32 + shift).rem_euclid(26);
        self.letters[new_index as usize]
      }
      None => c, // Non-alphabet chars get returned
    }
  }

  pub fn encode(&self, input: &str) -> String {
    let input_chars: Vec<char> = input.chars().collect();
    let key_len: usize = self.mask.len();  // key length comes from mask
    let mut output: String = String::new();
    let mut key_index: usize = 0;
    let mut idx = 0;
  
    for ch in &input_chars {
      if ! ch.is_whitespace() {
        // get row from matrix based on key character
        let first_idx = self.get_index(self.mask[key_index]).unwrap_or(0) as usize;
        let second_idx = self.get_index(input_chars[idx]).unwrap_or(0) as usize;
        output.push(self.matrix[first_idx][second_idx]);
      } else {
        output.push(' ');
      }
      key_index = (key_index + 1) % key_len;
      idx += 1;
    }
    output
  }

  #[allow(unused_variables)]
  pub fn decode(&self, input: &str) -> String {
    let output: &str = "HAPPY BIRTHDAY";
    output.to_string()
  }

  pub fn display_matrix(&self) {
    print!("[ ]  "); 
    for c in &self.letters {
        print!("[{}] ", c);
    }
    println!();

    for (i, row) in self.matrix.iter().enumerate() {
        print!("[{}] ", self.letters[i]);
        for ch in row {
            print!("| {} ", ch);
        }
        println!("|");
    }
  }

  pub fn inline_decode(&self, input: &str) -> String {
    let mut output = String::new();
    let mut key_index = 0;

    for c in input.to_uppercase().chars() {
      if c == ' ' { output.push(' '); }
      else {
        let shift = self.get_index(self.mask[key_index]).unwrap_or(0) as i32;
        let base = self.get_index(c).unwrap_or(0) as i32;
        let new_index = (base - shift).rem_euclid(26);
        output.push(self.letters[new_index as usize]);
        key_index = (key_index + 1) % self.mask.len();
      }
    }
    output
  }

  pub fn inline_encode(&self, input: &str) -> String {
    let mut output = String::new();
    let mut key_index = 0;

    for c in input.to_uppercase().chars() {
      if c == ' ' { output.push(' '); }
      else {
        let shift = self.get_index(self.mask[key_index]).unwrap_or(0) as i32;
        let base = self.get_index(c).unwrap_or(0) as i32;
        let new_index = (base + shift).rem_euclid(26);
        output.push(self.letters[new_index as usize]);
        key_index = (key_index + 1) %  self.mask.len();
      }
    }
    output
  }

  /// Performs the tradititional rotational detransformation
  pub fn rot_decode(&self, input: &str) -> String {
    input
    .to_uppercase()
    .chars()
    .map(|c| self.shift_char(c, -self.shift_key))
    .collect()
  }
  
  /// Performs the traditional rotational transformation
  pub fn rot_encode(&self, input: &str) -> String {
    input
      .to_uppercase()
      .chars()
      .map(|c| self.shift_char(c, self.shift_key))
      .collect()
  }
}

/// Generate cipher matrix
fn generate_matrix(start: &char) -> Vec<Vec<char>> {
  let mut output: Vec<Vec<char>> = Vec::with_capacity(26);
  let state: Vec<char> = rotated_alphabet(start);
  for c in state {
    output.push(rotated_alphabet(&c));
  }
  output
}

fn rotated_alphabet(start: &char) -> Vec<char> {
  let letters: Vec<char> = ('A'..='Z').collect();
  let start_index = letters.iter().position(|c| c == start).unwrap_or(0) as usize;
  // println!("Start Index: {}", start_index);
  letters.iter()
    .cycle()
    .skip(start_index)
    .take(26)
    .cloned()
    .collect()
}

fn main() {
  let code = 'H';
  let salt = "BABBAGE";
  let msg = "HAPPY BIRTHDAY";
  let cipher = Ceasar::new(code, salt);
  let encoded = cipher.encode(msg);
  let decoded = cipher.decode(&encoded);
  
  // cipher.display_matrix();
  println!("Input:     {}", msg);
  println!("Encrypted: {}", encoded);
  println!("Decrypted: {}", decoded);
}